File: src/net/codechunk/speedofsound/util/BluetoothDevicePreference.java
package net.codechunk.speedofsound.util;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothClass;
import android.bluetooth.BluetoothDevice;
import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.preference.DialogPreference;
import android.preference.PreferenceManager;
import android.util.AttributeSet;
import android.util.Log;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import net.codechunk.speedofsound.R;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * A preference dialog listing saved and paired Bluetooth devices.
 */
public class BluetoothDevicePreference extends DialogPreference {
    private static final String TAG = "BluetoothPreference";

    public static final String KEY = "enable_bluetooth_devices";

    protected Set<String> value;

    private View view;
    private List<PrettyBluetoothDevice> adapterDevices = new ArrayList<>();

    public BluetoothDevicePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    protected void onSetInitialValue(boolean restorePersistedValue, Object defaultValue) {
        if (restorePersistedValue) {
            this.value = getPersistedDevices();
        } else {
            this.value = new HashSet<>();
            persistDevices(this.value);
        }
    }

    @Override
    protected View onCreateDialogView() {
        super.onCreateDialogView();

        // reload persisted values (onSetInitialValue only called for first init)
        this.value = getPersistedDevices();

        // inflate it all
        LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.view = inflater.inflate(R.layout.bluetooth_preference_dialog, null);

        // get our paired music bluetooth devices
        this.adapterDevices.clear();
        BluetoothAdapter adapter = BluetoothAdapter.getDefaultAdapter();
        if (adapter != null) {
            Set<BluetoothDevice> deviceSet = adapter.getBondedDevices();
            for (BluetoothDevice bluetoothDevice : deviceSet) {
                BluetoothClass bluetoothClass = bluetoothDevice.getBluetoothClass();
                if (bluetoothClass.getMajorDeviceClass() == BluetoothClass.Device.Major.AUDIO_VIDEO) {
                    this.adapterDevices.add(new PrettyBluetoothDevice(bluetoothDevice));
                } else {
                    Log.d(TAG, "Skipping " + bluetoothDevice.getName() + " with class " + bluetoothClass.getDeviceClass());
                }
            }
        }

        // add them to the list
        ListView list = (ListView) this.view.findViewById(R.id.bluetooth_preference_listview);
        list.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);
        ArrayAdapter<PrettyBluetoothDevice> listAdapter =
                new ArrayAdapter<>(getContext(), android.R.layout.simple_list_item_multiple_choice, this.adapterDevices);
        list.setAdapter(listAdapter);
        setCheckedDevices(this.value);

        return this.view;
    }

    @Override
    public void onDialogClosed(boolean positiveResult) {
        super.onDialogClosed(positiveResult);

        if (!positiveResult) {
            return;
        }

        if (shouldPersist()) {
            persistDevices(getCheckedDevices());
        }

        notifyChanged();
    }

    private Set<String> getPersistedDevices() {
        return PreferenceManager.getDefaultSharedPreferences(getContext())
                .getStringSet(getKey(), new HashSet<String>());
    }

    private void persistDevices(Set<String> items) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
        SharedPreferences.Editor editor = prefs.edit();
        editor.putStringSet(getKey(), items);
        editor.apply();
    }

    /**
     * Get a set of checked devices from this dialog's ListView.
     */
    private Set<String> getCheckedDevices() {
        ListView list = (ListView) this.view.findViewById(R.id.bluetooth_preference_listview);
        Set<String> devices = new HashSet<>();

        SparseBooleanArray checked = list.getCheckedItemPositions();
        int size = list.getAdapter().getCount();
        for (int i = 0; i < size; i++) {
            if (!checked.get(i)) {
                continue;
            }

            PrettyBluetoothDevice device = this.adapterDevices.get(i);
            devices.add(device.getAddress());
        }

        return devices;
    }

    /**
     * Set the checked devices in the displayed ListView.
     */
    private void setCheckedDevices(Set<String> devices) {
        ListView list = (ListView) this.view.findViewById(R.id.bluetooth_preference_listview);

        int size = list.getAdapter().getCount();
        for (int i = 0; i < size; i++) {
            if (devices.contains(this.adapterDevices.get(i).getAddress())) {
                list.setItemChecked(i, true);
            }
        }
    }

    /**
     * Bluetooth device with a nice toString().
     */
    private class PrettyBluetoothDevice {
        private BluetoothDevice device;

        PrettyBluetoothDevice(BluetoothDevice device) {
            this.device = device;
        }

        @Override
        public String toString() {
            return this.device.getName();
        }

        public String getAddress() {
            return this.device.getAddress();
        }
    }

}

File: src/net/codechunk/speedofsound/util/BluetoothDevicePreference.kt
package net.codechunk.speedofsound.util

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.content.Context
import android.preference.DialogPreference
import android.preference.PreferenceManager
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.widget.ArrayAdapter
import android.widget.ListView
import net.codechunk.speedofsound.R
import java.util.*

/**
 * A preference dialog listing saved and paired Bluetooth devices.
 */
class BluetoothDevicePreference(context: Context, attrs: AttributeSet) : DialogPreference(context, attrs) {

    private var value: Set<String>? = null

    private var view: View? = null
    private val adapterDevices = ArrayList<PrettyBluetoothDevice>()

    private val persistedDevices: Set<String>
        get() = PreferenceManager.getDefaultSharedPreferences(context)
            .getStringSet(key, HashSet())!!

    /**
     * Get a set of checked devices from this dialog's ListView.
     */
    /**
     * Set the checked devices in the displayed ListView.
     */
    private var checkedDevices: Set<String>
        get() {
            val list = this.view!!.findViewById<View>(R.id.bluetooth_preference_listview) as ListView
            val devices = HashSet<String>()

            val checked = list.checkedItemPositions
            val size = list.adapter.count
            for (i in 0 until size) {
                if (!checked.get(i)) {
                    continue
                }

                val device = this.adapterDevices[i]
                devices.add(device.address)
            }

            return devices
        }
        set(devices) {
            val list = this.view!!.findViewById<View>(R.id.bluetooth_preference_listview) as ListView

            val size = list.adapter.count
            for (i in 0 until size) {
                if (devices.contains(this.adapterDevices[i].address)) {
                    list.setItemChecked(i, true)
                }
            }
        }

    override fun onSetInitialValue(restorePersistedValue: Boolean, defaultValue: Any?) {
        if (restorePersistedValue) {
            this.value = persistedDevices
        } else {
            this.value = HashSet()
            persistDevices(this.value!!)
        }
    }

    override fun onCreateDialogView(): View {
        super.onCreateDialogView()

        // reload persisted values (onSetInitialValue only called for first init)
        this.value = persistedDevices

        // inflate it all
        val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        this.view = inflater.inflate(R.layout.bluetooth_preference_dialog, null)

        // get our paired music bluetooth devices
        this.adapterDevices.clear()
        val adapter = BluetoothAdapter.getDefaultAdapter()
        if (adapter != null) {
            val deviceSet = adapter.bondedDevices
            for (bluetoothDevice in deviceSet) {
                this.adapterDevices.add(PrettyBluetoothDevice(bluetoothDevice))
            }
        }

        // add them to the list
        val list = this.view!!.findViewById<View>(R.id.bluetooth_preference_listview) as ListView
        list.choiceMode = ListView.CHOICE_MODE_MULTIPLE
        val listAdapter = ArrayAdapter(context, android.R.layout.simple_list_item_multiple_choice, this.adapterDevices)
        list.adapter = listAdapter
        checkedDevices = this.value!!

        return this.view!!
    }

    public override fun onDialogClosed(positiveResult: Boolean) {
        super.onDialogClosed(positiveResult)

        if (!positiveResult) {
            return
        }

        if (shouldPersist()) {
            persistDevices(checkedDevices)
        }

        notifyChanged()
    }

    private fun persistDevices(items: Set<String>) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        val editor = prefs.edit()
        editor.putStringSet(key, items)
        editor.apply()
    }

    /**
     * Bluetooth device with a nice toString().
     */
    private inner class PrettyBluetoothDevice internal constructor(private val device: BluetoothDevice) {

        val address: String
            get() = this.device.address

        override fun toString(): String {
            return this.device.name
        }
    }

    companion object {

        const val KEY = "enable_bluetooth_devices"
    }

}
